blocks/graph.py

```diff
@@ -319,3 +320,84 @@ def apply_noise(graph, variables, level, seed=None):
         replace[variable] = (variable +
                              rng.normal(variable.shape, std=level))
     return graph.replace(replace)
+
+
+def collect_parameters(graph, params):
+    """Replace parameters with a single shared variable.
+
+    This can be useful if you need to calculate the full Hessian of a
+    computational graph. It replaces parameters with slices of a single
+    large vectors like
+
+    >>> from blocks.utils import shared_floatx
+    >>> W1 = shared_floatx(numpy.random.rand(10, 10))
+    >>> W2 = shared_floatx(numpy.random.rand(10, 10))
+    >>> all_params = shared_floatx(numpy.concatenate(
+    ...     [W1.get_value().flatten(), W2.get_value().flatten()]))
+    >>> W1 = all_params[:W1.size]
+    >>> W2 = all_params[W1.size:]
+
+    Parameters
+    ----------
+    graph : :class:`ComputationGraph` instance
+        The managed Theano graph in which to collect parameters.
+    params : list of Theano shared variables
+        The parameters whose values should be collected.
+
+    Returns
+    -------
+    ComputationGraph instance
+        A new Theano graph which has all the given parameters collected
+        into a single large shared variable.
+
+    Notes
+    -----
+    Note that this replacement makes the training of the model
+    significantly slower because of the large amount of Theano's
+    ``set_subtensor`` calls needed to train the model.
+
+    Examples
+    --------
+    >>> from blocks.bricks import MLP, Sigmoid
+    >>> from blocks.bricks.cost import SquaredError
+    >>> from theano import tensor
+    >>> x = tensor.matrix()
+    >>> mlp = MLP(activations=[Sigmoid(), Sigmoid()], dims=[784, 100, 784])
+    >>> cost = SquaredError().apply(x, mlp.apply(x))
+    >>> cg = ComputationGraph(cost)
+    >>> new_cg = collect_parameters(cg, cg.shared_variables)
+
+    The new graph only has a single shared variable.
+
+    >>> new_cg.shared_variables
+    [collected_params]
+
+    The bricks' variables have been replaced with reshaped segments of this
+    single shared variable.
+
+    >>> from blocks.filter import VariableFilter
+    >>> from blocks.roles import PARAMETER
+    >>> var_filter = VariableFilter(roles=[PARAMETER])
+    >>> var_filter(new_cg.variables) # doctest: +SKIP
+    [Reshape{1}.0, Reshape{1}.0, Reshape{2}.0, Reshape{2}.0]
+
+    """
+    param_values, param_sizes, param_shapes = [], [], []
+    for param in params:
+        param_values.append(param.get_value(borrow=True))
+        param_sizes.append(param_values[-1].size)
+        param_shapes.append(param_values[-1].shape)
+
+    new_params = shared_floatx_zeros(sum(param_sizes))
+    new_params.set_value(numpy.concatenate([value.flatten()
+                                            for value in param_values]))
+    new_params.name = 'collected_params'
```

Comment:
Okay, for identifying which shared variables are there to model randomness we can probably ask for support from Theano developers. But I have just remembered that the learning rate will very soon become a shared variable as well. One can mark it with a role `NOT_PARAMETER`, but I feel bad about it.

And I fact I should say I do not like the idea that a non-shared variable can be a parameter. This conflicts with the standard semantic of the word: parameter is something external to the formula, something independent. We can introduce another role `PARAMETER_SUBSTITUTE` for the cases when a parameter is replaced by a function of other parameters. Or we can use a combination of two roles `PARAMETER` and `SUBSTITUTE` for these cases. 

The one-liner `apply_noise(new_cg, VariableFilter(role=[WEIGHTS])(new_cg.variables))` becomes by the way extremely dangerous when you work with an autoencoder, because noise will be added twice, both to the encoder weights and to the decoder weights obtained by transposition of the former. And to be honest, it is not going to be easy to add different noise to these two matrices...